lib/documents.js

```diff
@@ -2263,12 +2268,184 @@ function writeAllDocumentsImpl(options) {
   let operation = new Operation(
       'read forestInfo', this.client, requestOptions, 'empty', 'single'
   );
-  operation.stream = inputStream;
-  let result = requester.startRequest(operation);
-  operation.response = result;
+
+  let jobState = {
+    docInstance : this,
+    stream: inputStream,
+    requesterCount: 0,
+    docsWrittenSuccessfully:0,
+    docsFailedToBeWritten: 0,
+    jobOptions: (jobOptions)? mlutil.copyProperties(jobOptions):{}
+  };
+
+ if(jobState.jobOptions){
+   // Check for jobOptions.concurrentRequests
+   if(jobState.jobOptions.concurrentRequests) {
+     const concurrentRequests = jobState.jobOptions.concurrentRequests;
+     if (concurrentRequests.multipleOf) {
+       if (!new Set(['forests', 'hosts']).has(String(concurrentRequests.multipleOf))) {
+         throw new Error('Invalid value for onCompletion.multipleOf. Value must be forests or hosts.');
+       }
+     } else {
+       concurrentRequests.multipleOf = 'forests';
+     }
+
+     if (concurrentRequests.multiplier && concurrentRequests.multiplier < 0) {
+       throw new Error('concurrentRequests.multiplier cannot be less than zero');
+     } else {
+       concurrentRequests.multiplier = 4;
+     }
+     jobState.jobOptions.concurrentRequests = concurrentRequests;
+   }
+   else {
+     jobState.jobOptions.concurrentRequests = {multipleOf:'forests', multiplier:4};
+   }
+
+   // Check for jobOptions.batchSize
+   jobState.jobOptions.batchSize = (jobState.jobOptions.batchSize && jobState.jobOptions.batchSize >0) ? jobState.jobOptions.batchSize:100;
+
+   //Check for defaultMetadata
+   jobState.jobOptions.defaultMetadata = (jobState.jobOptions.defaultMetadata)? jobState.jobOptions.defaultMetadata:{};
```

Comment:
Defaulting to an empty metadata object results in sending an empty metadata payload to the server, which adds to the expense of the request without benefit to the application. 

Would it be better to default to null and skip sending the metadata payload if it is null?

Reply:
Thanks Erik.! Changed the same for "transform" as well.